### PR TITLE
Make <&str as Arbitrary>::arbitrary_take_rest less likely to fail

### PR DESCRIPTION
**Pull request by Ekleog-NEAR**
Before the changes, if any character in the remaining data were to be invalid utf8, it would fail to generate a string.
After the change, it takes all it can up to the first invalid utf8 byte.

### CODE REVIEW

**Review by fitzgen — COMMENTED** [review]
Can you add a test for this new behavior?

File: src/lib.rs [hunk 1]
```diff
@@ -834,29 +834,33 @@ where
...
+    match str::from_utf8(u.peek_bytes(size).unwrap()) {
+        Ok(s) => {
+            u.bytes(size).unwrap();
+            Ok(s)
+        }
+        Err(e) => {
+            let i = e.valid_up_to();
+            let valid = u.bytes(i).unwrap();
+            let s = unsafe {
+                debug_assert!(str::from_utf8(valid).is_ok());
```
Comment (Manishearth) [inline src/lib.rs:847]: thought: should we be including this? fuzzing is usually run with debug assertions on so the debug_assert doesn't get us much. I am comfortable with more thoroughly documenting the safety situation instead

**Review by Manishearth — APPROVED** [review]

File: src/lib.rs (hunk 1 above)
Comment (fitzgen) [reply src/lib.rs:847]: We have a couple other debug asserts that also maybe should only be enabled when running the `arbitrary` crate tests. We can figure out what to do with all of them in a follow up, no need to hold this PR up.

Comment (Ekleog-NEAR) [reply src/lib.rs:847]: This `debug_assert` comes from the previous implementation of `arbitrary`; if you think it’d be better to replace it with a comment that `valid_up_to` can only generate valid strings, or with another macro that’d assert only on eg. an `internal-tests` arbitrary feature, I’d be happy to make a follow-up PR :)

**Pull request merged**